pkg/runtime/handler.go

```diff
@@ -101,19 +101,45 @@ func (h *handlerFactory) Build(handler *pb.Handler, instances []*pb.Instance, en
 	hndlrBldrInfo, _ := h.builderInfoFinder(handler.Adapter)
 
 	hndlrBldr := hndlrBldrInfo.NewBuilder()
+
 	if hndlrBldr == nil {
 		msg := fmt.Sprintf("nil HandlerBuilder instantiated for adapter '%s' in handler config '%s'", handler.Adapter, handler.Name)
 		glog.Warning(msg)
 		return nil, errors.New(msg)
 	}
 
+	// validate if the builder supports all the necessary interfaces
+	for _, tmplName := range hndlrBldrInfo.SupportedTemplates {
+		// ti should be there for a valid configuration.
+		ti, _ := h.tmplRepo.GetTemplateInfo(tmplName)
+		if supports := ti.BuilderSupportsTemplate(hndlrBldr); !supports {
+			// adapter's builder is bad since it does not support the necessary interface
+			msg := fmt.Sprintf("adapter is invalid because it does not implement interface '%s'. "+
+				"Therefore, it cannot support template '%s'", ti.BldrInterfaceName, tmplName)
+			glog.Error(msg)
+			return nil, fmt.Errorf(msg)
+		}
+	}
+
 	var hndlr adapter.Handler
 	hndlr, err = h.build(hndlrBldr, infrdTypsByTmpl, handler.Params, env)
 	if err != nil {
 		msg := fmt.Sprintf("cannot configure adapter '%s' in handler config '%s': %v", handler.Adapter, handler.Name, err)
 		glog.Warning(msg)
 		return nil, errors.New(msg)
 	}
+	// validate if the handler supports all the necessary interfaces
+	for _, tmplName := range hndlrBldrInfo.SupportedTemplates {
+		// ti should be there for a valid configuration.
+		ti, _ := h.tmplRepo.GetTemplateInfo(tmplName)
+		if supports := ti.HandlerSupportsTemplate(hndlr); !supports {
+			// adapter is bad since it does not support the necessary interface
```

Comment:
Nit: we cannot change it in this PR, but
ti.handlerSupportsTemplate can just be Suppprts

Reply:
Yeah... later.